### PR TITLE
fix: include error code in RpcLogger

### LINKED ISSUE #1129: Feature Request: have `jsonrpsee::server::logger` return result as RpcResult instead of a string?
I have some prometheus metrics set up on a logger that implements `jsonrpsee::server::logger`. All errors currently increment a single `errors_by_route` metric, but now I'd like to increment different error metrics depending on the response error code. However, the logger only provides a `success: bool` from `on_result`, or a `result: &str` from `on_response`. Could we have `on_response` return the actual RpcResult instead?

### PR DESCRIPTION
**Pull request by niklasad1**
Close #1129 

In addition it contains a bug fix that were introduced by the `IntoResponse` trait and I forgot to check the outcome from the `ResponsePayload` where the user could inject there own JSON-RPC error codes etc.

Thanks to @wlmyng for spotting the bug.

### CODE REVIEW

File: tests/tests/metrics.rs [hunk 1]
```diff
@@ -154,12 +159,16 @@ async fn ws_server_logger() {
 	let res: Result<String, Error> = client.request("unknown_method", rpc_params![]).await;
 	assert!(res.is_err());
 
+	let res: Result<String, Error> = client.request("err", rpc_params![]).await;
```
Comment (niklasad1) [inline tests/tests/metrics.rs:162]: Expanded to test for failed RPC calls too

**Comment by wlmyng** [comment]
Thank you for looking into this!

File: core/src/server/helpers.rs [hunk 2]
```diff
@@ -224,7 +274,7 @@ impl MethodResponse {
 	pub fn error<'a>(id: Id, err: impl Into<ErrorObject<'a>>) -> Self {
 		let err = ResponsePayload::error_borrowed(err);
 		let result = serde_json::to_string(&Response::new(err, id)).expect("JSON serialization infallible; qed");
-		Self { result, success: false }
+		Self { result, success_or_error: MethodResponseResult::Failed(ErrorCode::InternalError.code()) }
```
Comment (jsdw) [inline core/src/server/helpers.rs:277]: maybe a dumb question but why does this set an error code which is different from the error code in the `ErrorObject`?

Comment (niklasad1) [reply core/src/server/helpers.rs:277]: it's a bug, good catch

**Review by jsdw — APPROVED** [review]
looks good!

File: core/src/server/helpers.rs [hunk 3]
```diff
@@ -173,13 +173,56 @@ pub fn prepare_error(data: &[u8]) -> (Id<'_>, ErrorCode) {
...
+		self.success_or_error.is_success()
+	}
+}
+
+/// Represent the outcome of a method call success or failed.
+#[derive(Debug, Copy, Clone)]
+pub enum MethodResponseResult {
+	/// The method call was successful.
+	Success,
+	/// The method call with error code.
```
Comment (niklasad1) [inline core/src/server/helpers.rs:202]:
```suggestion
	/// The method call failed with error code.
```

**Review by lexnv — APPROVED** [review]
LGTM!

**Pull request merged**